DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,866 @@
+method from the `TestsUtils` class.
+
+```javascript
+it('waits for no locks', async () => {
+  const datasetResponse = await DatasetHelper.create()
+
+  await DatasetHelper.publish(datasetResponse.persistentId)
+  await TestsUtils.waitForNoLocks(datasetResponse.persistentId)
+
+  await datasetRepository.getByPersistentId(datasetResponse.persistentId).then((dataset) => {
+    if (!dataset) {
+      throw new Error('Dataset not found')
+    }
+    expect(dataset.locks).to.deep.equal([])
+  })
+})
+```
+
+</details>
+
+### 3. **End-to-End (e2e) Tests:**
+
+End-to-end tests simulate real user scenarios, covering the complete flow of the application:
+
+- **User Workflows:** Focus on common user paths, like searching for a file, logging in, or creating a Dataset. Ensure
+  these workflows work from start to finish.
+- **Avoid Redundancy:** Do not replicate tests covered by unit tests. E2E tests should cover broader user experiences.
+  It is important to note that e2e tests are slower and more brittle than unit tests, so we use them sparingly.
+- **Tools and Frameworks:** We use Cypress for e2e tests, to test the application's behavior from the user's perspective.
+  It allows us to simulate user interactions and test the application's behavior in a real-world scenario.
+
+<details>
+<summary><strong>Example of an E2E Test</strong></summary>
+
+```javascript
+//tests/e2e-integration/e2e/sections/create-dataset/CreateDatasetForm.spec.tsx
+
+describe('Create Dataset', () => {
+  before(() => {
+    TestsUtils.setup()
+  })
+  beforeEach(() => {
+    TestsUtils.login()
+  })
+
+  it('navigates to the new dataset after submitting a valid form', () => {
+    cy.visit('/spa/datasets/create')
+
+    cy.findByLabelText(/Title/i).type('Test Dataset Title')
+    cy.findByLabelText(/Author Name/i).type('Test author name', { force: true })
+    cy.findByLabelText(/Point of Contact E-mail/i).type('email@test.com')
+    cy.findByLabelText(/Description Text/i).type('Test description text')
+    cy.findByLabelText(/Arts and Humanities/i).check()
+
+    cy.findByText(/Save Dataset/i).click()
+
+    cy.findByRole('heading', { name: 'Test Dataset Title' }).should('exist')
+    cy.findByText(DatasetLabelValue.DRAFT).should('exist')
+    cy.findByText(DatasetLabelValue.UNPUBLISHED).should('exist')
+  })
+})
+```
+
+</details>
+
+> **Note:** The current e2e tests are failing due to the following reasons:
+>
+> - **Dataset JSDataverse Repository -> gets the total dataset count**: Calling `destroyAll()` before the "gets the total
+> - dataset count" test in `DatasetJSDataverseRepository.spec.ts` causes an optimistic lock exception in Dataverse.
+>
+> **Solution:** We need to either reset the database after each test or find an alternative method to avoid the optimistic
+> lock exception. Check the issue [here](https://github.com/IQSS/dataverse-frontend/issues/294).
+
```

Comment:
This issue has been QA'd and should be merged soon, but a new issue about the local environment is here https://github.com/IQSS/dataverse-frontend/issues/371